sql/migrations/0024-device-roles.sql

```diff
@@ -0,0 +1,35 @@
+SELECT run_migration(24, $$
+
+create table device_service (
+	id   uuid primary key default gen_random_uuid(),
+	name text not null unique,
+
+	created     timestamptz not null default current_timestamp,
+	updated     timestamptz not null default current_timestamp
+);
+
+
+create table device_role (
+	id                  uuid primary key default gen_random_uuid(),
+	name                text not null,
+	description         text,
+	hardware_product_id uuid not null references hardware_product(id),
+
+	created     timestamptz not null default current_timestamp,
+	updated     timestamptz not null default current_timestamp,
+	deactivated timestamptz
+);
+create unique index on device_role(name) where deactivated is null;
```

Comment:
How will this handle instances where we have headnodes or admin servers that might be on R510s or R730s or whatever, depending on the datacenter?

Should this be an index between name and hardware_product, so we can account for

compute_node hw1
compute_node hw2
headnode hw1
headnode hw3
manta_storage shrimp_v3
manta_storage shrimp_v4

Or am I misunderstanding?

Reply:
The 'name' field is intended for use in the UIs and tools to present a list of roles. It has no internal semantic meaning. "headnode on a r510" is a role where the hardware product is "r510" and the service list contains an entry for "headnode".

I'm going to drop the name field entirely, I think, to prevent folks from thinking that creating a role with a text name of "triton - r730" makes it so. In the tools, we'll use the description field and build "names" using a combination of the hardware profile and device service name fields.

Reply:
Done.